File: C++/1/002-AddTwoNumbers.cc
// Definition for singly-linked list.
#include "leetcode.h"
// #include "linked-list.h"
 
class Solution {
 public:
  ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
    if (l1 == NULL) {
      return l2;
    }
    if (l2 == NULL) {
      return l1;
    }
    ListNode* ret = l1;
    int carry = 0;
    while (true) {
      l1->val += l2->val + carry;
      if (l1->val > 9) {
        carry = 1;
        l1->val -= 10;
      } else {
        carry = 0;
      }
      if (l1->next == NULL) {
        l1->next = l2->next;  // L1 is finished. Move l2 to l1 to continue.
        break;
      } else if (l2->next == NULL) {
        break;
      }
      l1 = l1->next;
      l2 = l2->next;
    }
    while (l1->next != NULL && carry) {  // Need to check the last node.
      l1 = l1->next;  // Aware!
      l1->val += carry;
      if (l1->val > 9) {
        carry = 1;
        l1->val -= 10;
      } else {
        return ret;
      }
    }
    if (carry) {  // Add a new node by the head of l2.
      l2->val = carry;
      l2->next = nullptr;
      l1->next = l2;
    }
    return ret;
  }

  ListNode* addTwoNumbersLeak(ListNode* l1, ListNode* l2) {
    // Define a dummy node for linked list operation.
    ListNode header(-1), *current = &header;
    int carry = 0;
    while (l1 || l2 || carry) {
      int sum = (l1 ? l1->val : 0) + (l2 ? l2->val : 0) + carry;
      current->next = new ListNode(sum % 10);  // Memory leaks!
      current = current->next;

      carry = sum / 10;
      l1 = l1 ? l1->next : nullptr;
      l2 = l2 ? l2->next : nullptr;
    }
    return header.next;
  }
};

int main(int argc, char** argv) {
  Solution solution;
  ListNode nodes[2] = {2, 0};

  ListNode* l1 = &nodes[0];
  // Print(l1);

  ListNode* l2 = &nodes[1];
  // l2->next = &nodes[2];
  // Print(l2);

  ListNode* root = solution.addTwoNumbers(l1, l2);
  Print(root);
  return 0;
}

File: Python-2/002-AddTwoNumbers.py
# Definition for singly-linked list.
class ListNode(object):
    def __init__(self, x, next=None):
        self.val = x
        self.next = next


class Solution(object):
    def addTwoNumbers(self, l1, l2):
        """
        :type l1: ListNode
        :type l2: ListNode
        :rtype: ListNode
        """
        head = ListNode(-1)
        curr = head
        carry = 0
        while l1 or l2 or carry:
            current_sum = (l1.val if l1 else 0) + (l2.val if l2 else 0) + carry
            curr.next = ListNode(current_sum % 10)
            carry = current_sum / 10

            curr = curr.next
            l1 = l1.next if l1 else None
            l2 = l2.next if l2 else None
        return head.next

if __name__ == "__main__":
    l1 = ListNode(1)
    l2 = ListNode(9, ListNode(9))
    l = Solution().addTwoNumbers(l1, l2)
    while l:
        print l.val,
        l = l.next
